File: backend/app/api/routes/catalog.py
from fastapi import APIRouter, Query

from app.data.repository import get_colleges, get_courses, get_filter_options

router = APIRouter(prefix="/catalog", tags=["Catalog"])


@router.get("/options")
def catalog_options(
    country: str | None = Query(default=None, min_length=2),
    stream: str | None = Query(default=None, min_length=2),
    level: str | None = Query(default=None, min_length=2),
) -> dict[str, list[str]]:
    normalized_level = level.upper() if level else None
    return get_filter_options(country=country, stream=stream, level=normalized_level)


@router.get("/colleges")
def catalog_colleges(
    country: str = Query(..., min_length=2),
    stream: str = Query(..., min_length=2),
    level: str = Query(..., min_length=2),
) -> dict[str, list[str]]:
    return {"colleges": get_colleges(country=country, stream=stream, level=level.upper())}


@router.get("/courses")
def catalog_courses(
    country: str = Query(..., min_length=2),
    stream: str = Query(..., min_length=2),
    level: str = Query(..., min_length=2),
    college_name: str = Query(..., min_length=2),
) -> dict[str, list[str]]:
    return {
        "courses": get_courses(
            country=country,
            stream=stream,
            level=level.upper(),
            college_name=college_name,
        )
    }

File: frontend/src/services/catalog.ts
import api from "./api";
import type { CatalogOptions } from "@/types/roi";

export async function getCatalogOptions(): Promise<CatalogOptions> {
  const { data } = await api.get<CatalogOptions>("/catalog/options");
  return data;
}

export async function getColleges(params: {
  country: string;
  stream: string;
  level: string;
}): Promise<string[]> {
  const { data } = await api.get<{ colleges: string[] }>("/catalog/colleges", { params });
  return data.colleges;
}

export async function getCourses(params: {
  country: string;
  stream: string;
  level: string;
  college_name: string;
}): Promise<string[]> {
  const { data } = await api.get<{ courses: string[] }>("/catalog/courses", { params });
  return data.courses;
}
